# Tools

Tools are code that can be executed by an agent to perform actions. A tool can be a simple function such as a calculator, or an API call to a third-party service such as stock price lookup or weather forecast. In the context of AI agents, tools are designed to be executed by agents in response to model-generated function calls.


There are three types of Tools :<br>
1. Custom Function Tools
2. Built-in Tools
3. Third Party Calling Tools (eg. Langchain built-in tools)

In [2]:
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from autogen_core.tools import FunctionTool

In [8]:
import asyncio, os

from dotenv import load_dotenv

# Load environment variables
load_dotenv()
api_key= os.getenv("OPENAI_API_KEY")

if not api_key:
    raise ValueError("Please set the OPENAI_API_KEY environment variable.")

In [9]:
model_client= OpenAIChatCompletionClient(model="gpt-4o", api_key=api_key)

In [10]:
def reverse_string(text: str) -> str:
    '''
    Reverse the given text

    input: str

    output: str

    The reverse string is returned.
    '''
    return text[::-1]

In [ ]:
# We can pass only one custom function within FunctionTool at a time
reverse_tool= FunctionTool(reverse_string, description="A tool to reverse a string")

In [12]:
reverse_tool

In [13]:
agent= AssistantAgent(
    name="ReverseStringAgent",
    model_client=model_client,
    system_message='You are a helpful assistant that can reverse string using reverse_string tool. Give the result with summary',
    tools=[reverse_tool] # Here, We can pass multiple tools
)

In [14]:
agent

In [15]:
response= await agent.run(task='"Reverse the string "Hello, World!"')

In [16]:
response

TaskResult(messages=[TextMessage(id='0e4ca9ef-2809-4029-9871-a343b56bd4f2', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 6, 13, 57, 24, 930401, tzinfo=datetime.timezone.utc), content='"Reverse the string "Hello, World!"', type='TextMessage'), ToolCallRequestEvent(id='4a4b5d3c-158a-4a57-b940-0b29968d46a7', source='ReverseStringAgent', models_usage=RequestUsage(prompt_tokens=80, completion_tokens=17), metadata={}, created_at=datetime.datetime(2025, 7, 6, 13, 57, 26, 527390, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_bMHNYASMzRdGhVHO4Aet6VzG', arguments='{"text":"Hello, World!"}', name='reverse_string')], type='ToolCallRequestEvent'), ToolCallExecutionEvent(id='477ab332-497b-4933-9a08-58f16825f00b', source='ReverseStringAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 6, 13, 57, 26, 528896, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='!dlroW ,olleH', name='reverse_string', c

TaskResult(
    messages=[
        TextMessage(id='0e4ca9ef-2809-4029-9871-a343b56bd4f2', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 6, 13, 57, 24, 930401, tzinfo=datetime.timezone.utc), content='"Reverse the string "Hello, World!"', type='TextMessage'), 
        
        ToolCallRequestEvent(id='4a4b5d3c-158a-4a57-b940-0b29968d46a7', source='ReverseStringAgent', models_usage=RequestUsage(prompt_tokens=80, completion_tokens=17), metadata={}, created_at=datetime.datetime(2025, 7, 6, 13, 57, 26, 527390, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_bMHNYASMzRdGhVHO4Aet6VzG', arguments='{"text":"Hello, World!"}', name='reverse_string')], type='ToolCallRequestEvent'), 
        
        ToolCallExecutionEvent(id='477ab332-497b-4933-9a08-58f16825f00b', source='ReverseStringAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 6, 13, 57, 26, 528896, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='!dlroW ,olleH', name='reverse_string', call_id='call_bMHNYASMzRdGhVHO4Aet6VzG', is_error=False)], type='ToolCallExecutionEvent'), 
        
        ToolCallSummaryMessage(id='d54694f1-b61a-4111-ba2d-a3b4df663037', source='ReverseStringAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 6, 13, 57, 26, 528896, tzinfo=datetime.timezone.utc), content='!dlroW ,olleH', type='ToolCallSummaryMessage', tool_calls=[FunctionCall(id='call_bMHNYASMzRdGhVHO4Aet6VzG', arguments='{"text":"Hello, World!"}', name='reverse_string')], results=[FunctionExecutionResult(content='!dlroW ,olleH', name='reverse_string', call_id='call_bMHNYASMzRdGhVHO4Aet6VzG', is_error=False)])
        ], stop_reason=None)

### Output Summary

TextMessage(content='"Reverse the string "Hello, World!"')

ToolCallRequestEvent(content=[FunctionCall(id='call_bMHNYASMzRdGhVHO4Aet6VzG', arguments='{"text":"Hello, World!"}', name='reverse_string')])

ToolCallExecutionEvent( content=[FunctionExecutionResult(content='!dlroW ,olleH', name='reverse_string', call_id='call_bMHNYASMzRdGhVHO4Aet6VzG', is_error=False)])

ToolCallSummaryMessage(tool_calls=[FunctionCall(id='call_bMHNYASMzRdGhVHO4Aet6VzG', arguments='{"text":"Hello, World!"}', name='reverse_string')], results=[FunctionExecutionResult(content='!dlroW ,olleH', name='reverse_string', call_id='call_bMHNYASMzRdGhVHO4Aet6VzG', is_error=False)])
